Sources/SwiftProtobuf/TextFormatEncoder.swift

```diff
@@ -14,165 +14,261 @@
+private let asciiMinus = UInt8(ascii: "-")
+private let asciiBackslash = UInt8(ascii: "\\")
+private let asciiDoubleQuote = UInt8(ascii: "\"")
+private let asciiZero = UInt8(ascii: "0")
+private let asciiOpenCurlyBracket = UInt8(ascii: "{")
+private let asciiCloseCurlyBracket = UInt8(ascii: "}")
+private let asciiOpenSquareBracket = UInt8(ascii: "[")
+private let asciiCloseSquareBracket = UInt8(ascii: "]")
+private let asciiNewline = UInt8(ascii: "\n")
+private let asciiLowerB = UInt8(ascii: "b")
+private let asciiLowerF = UInt8(ascii: "f")
+private let asciiLowerN = UInt8(ascii: "n")
+private let asciiLowerR = UInt8(ascii: "r")
+private let asciiLowerT = UInt8(ascii: "t")
+private let asciiLowerV = UInt8(ascii: "v")
+
 /// TextFormatEncoder has no public members.
-public final class TextFormatEncoder {
-    var text: String = ""
+internal struct TextFormatEncoder {
+    private var data = [UInt8]()
     private var tabLevel = 0
-    var result: String { return text }
+    var stringResult: String {
+        get {
+            return String(bytes: data, encoding: String.Encoding.utf8)!
+        }
+    }
+
+    private mutating func append(text: String) {
+        data.append(contentsOf: text.utf8)
+    }
 
     init() {}
 
+    private mutating func appendFieldName(name: StaticString, inExtension: Bool) {
+        for _ in 0..<tabLevel {
+            data.append(asciiSpace)
+            data.append(asciiSpace)
+        }
+        if inExtension {
+            data.append(asciiOpenSquareBracket)
+        }
+        let buff = UnsafeBufferPointer(start: name.utf8Start, count: name.utf8CodeUnitCount)
+        data.append(contentsOf: buff)
+        if inExtension {
+            data.append(asciiCloseSquareBracket)
+        }
+    }
+
     // In Text format, fields with simple types write the name with
     // a trailing colon:
     //    name_of_field: value
-    func startField(name: String) {
-        for _ in 0..<tabLevel {
-            text.append("  ")
-        }
-        text.append(name)
-        text.append(":")
-        text.append(" ")
+    mutating func startField(name: StaticString, inExtension: Bool) {
+        appendFieldName(name: name, inExtension: inExtension)
+        data.append(asciiColon)
+        data.append(asciiSpace)
     }
 
     // In Text format, a message-valued field writes the name
     // without a trailing colon:
     //    name_of_field {key: value key2: value2}
-    func startMessageField(name: String) {
-        for _ in 0..<tabLevel {
-            text.append("  ")
-        }
-        text.append(name)
-        text.append(" ")
+    mutating func startMessageField(name: StaticString, inExtension: Bool) {
+        appendFieldName(name: name, inExtension: inExtension)
+        data.append(asciiSpace)
     }
 
-    func endField() {
-        text.append("\n")
+    mutating func endField() {
+        data.append(asciiNewline)
     }
 
-    func startObject() {
+    mutating func startObject() {
         tabLevel += 1
```

Comment:
So if you keep a var currentIndent = "" instead of a tabLevel, you can simply add two spaces when increasing and truncate two when decreasing.  Then the indent is always a single append instead of a loop with two appends per cycle.